### ISSUE TITLE
Very high cpu usage

### ISSUE THREAD

**Issue by fushugaku** [zenangst/KeyboardCowboy#565]
Hi, after the update the cpu utilisation is very high.
version 3.25.0
<img width="936" alt="image" src="https://github.com/user-attachments/assets/30500500-8b93-401a-8271-6c0fea07d253">
 
[Sample of Keyboard Cowboy.txt](https://github.com/user-attachments/files/17571465/Sample.of.Keyboard.Cowboy.txt)

**Comment by zenangst** [comment]
Hey @fushugaku, would you mind sharing what kind of commands you are running in your setup?
Also, which version of macOS are you using?

**Comment by fushugaku** [comment]
same, after restarting everything went back to normal.
I use shortcuts like 
fn + btn  next_btn, nothing exotic, just opening apps

**Comment by zenangst** [comment]
I think this might be a SwiftUI issue, not the core logic of the app. I’ll make sure to fix it first.

I’ve seen some spikes in the past, but I haven’t been able to pinpoint which part of the app is causing them.

Thanks for bringing this to my attention ❤️

**Comment by zenangst** [comment]
I did push out another version today which includes some compatibility fixes for older version of macOS. Did any of you have time to try that out to see if that also misbehaves?

ping @fushugaku @bassamsdata

**Comment by bassamsdata** [comment]
Thank you. I just updated the app, I'll try it out and report if the issue happened again.

**Comment by zenangst** [comment]
@bassamsdata @fushugaku
Did the latest update help at all?

### Small developer update.

Here's the deal.

SwiftUI keeps the window around even after it's closed, which is not ideal for memory usage of Keyboard Cowboy. I think it's a design choice for apps that aren't document-based.

So, I've decided to rewrite all the window handling in Keyboard Cowboy. Instead of using SwiftUI, I'll go for an AppKit-based approach.

Right now, most of the app is written in SwiftUI, but we're thinking about switching back to using `NSWindow`s that use SwiftUI as their content views.

The main reason for this is that we can improve how we handle windows. We'll have more control over each window and their lifecycle, and the app can decide when to release the window from memory. This means we can use less memory when the app is running in the background.

I had to think really hard about this, but I finally came up with a solution. One thing that helped me was @MrKai77's framework [DynamicNotchKit](https://github.com/MrKai77/DynamicNotchKit). It's basically the same idea.

If he hadn't shared his open-source work, this solution might have taken a lot longer to find. Thanks to him, I was able to get it done!

Here's an update on the rewrite:

I've converted all the windows, but they still need to be configured and tested to make sure the user experience is the same or better than before.

In short, the rewrite is underway to improve control and reduce memory usage.

**Comment by fushugaku** [comment]
@zenangst, hello, I guess it did help, I haven't experienced cpu spikes after the fix.
Thank you, KeyboardCowboy is awesome :)

**Comment by bassamsdata** [comment]
Thank you for all the information and for your hard work on this fantastic app!

I haven’t noticed any spike in CPU usage yet, though I haven’t been on my computer much recently.

**Comment by zenangst** [comment]
I've been making great progress on this task, and I'm currently testing it myself. I've been able to fix some of the CPU-intensive tasks and reduce the memory footprint when any Keyboard Cowboy windows are opened or closed.

And the best part? The UX has been improved because we now have more granular control.

Thanks again for reporting this. Internal window handling for Keyboard Cowboy was overdue, and this rewrite is really taking the application to the next level.

I still need to test this against older versions of macOS before it can be released.

**Comment by zenangst** [comment]
foo

**Comment by bassamsdata** [comment]
Thank you for all the hard work you've put into this issue. 

For some reason, the app has started spiking CPU usage today. I’m unsure if it's due to a conflict with other programs, but when it spikes, it continues until I restart the app. I'll investigate further tomorrow to see if that's the case.

Thank you again.

**Comment by zenangst** [comment]
Hey @bassamsdata, this is one of the main reasons I'm working on the window rewrite. I want to make sure that nothing interferes with in the background after the UI has been dismissed. That's the theory, anyway.

**Comment by zenangst** [comment]
@bassamsdata @fushugaku If I make a new version and give you a link here, would you mind taking that for a spin and helping beta-testing it? It would help a lot to fast-track the release.

**Comment by fushugaku** [comment]
yes, I would like to beta test

**Comment by zenangst** [comment]
Here's a sneak peek (almost ready for the world) version.

[Keyboard Cowboy 3.25.2.dmg.zip](https://github.com/user-attachments/files/17618125/Keyboard.Cowboy.3.25.2.dmg.zip)

Thanks again for taking the time to test the application.

**Comment by zenangst** [comment]
@fushugaku @bassamsdata did you guys encounter any oddities in the test version?

**Comment by fushugaku** [comment]
no oddities yet (ran for about an hour) :)
I'll send a comment if I find any

**Comment by zenangst** [comment]
@fushugaku thanks again for beta-testing.
I'm giving you a shoutout in the release notes for the upcoming version 😎

**Comment by fushugaku** [comment]
don't know if related, I decided to make "code" group today, to map shortcuts for vscode, and the app stuck on this screen, and nor Ok, nor Cancel buttons work.
Added two samples where i perform multiple clicks on the buttons
- ok is where i click okay button
- cancel when i click cancel button
<img width="594" alt="image" src="https://github.com/user-attachments/assets/a2143
[Sample of Keyboard Cowboy press OK.txt](https://github.com/user-attachments/files/17634510/Sample.of.Keyboard.Cowboy.press.OK.txt)

[Sample of Keyboard Cowboy press cancel.txt](https://github.com/user-attachments/files/17634509/Sample.of.Keyboard.Cowboy.press.cancel.txt)
95c-db5b-4c48-ac7e-21aa071d8f43">

**Comment by bassamsdata** [comment]
I've been testing the app with my usual workflow since yesterday, and I haven't noticed any unusual CPU activity—everything seems quiet and normal.

I also tested some shortcuts and didn't encounter any issues, except for the same one fushugaku mentioned, which happened for me as well.

**Comment by fushugaku** [comment]
so far so good, the bug with groups was fixed

**Comment by zenangst** [comment]
Gentlemen, I think we have a new version on our hands 😎

**Comment by bassamsdata** [comment]
It looks good to me so far! Thanks for the hardwork
